ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java

```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
     if (usedCapacity >= (storeConfig.storeMinUsedCapacityToTriggerCompactionInPercentage / 100.0) * totalCapacity) {
       List<String> potentialLogSegments = blobStore.getLogSegmentsNotInJournal();
       if (potentialLogSegments != null) {
-        details = new CompactionDetails(time.milliseconds() - messageRetentionTimeInMs, potentialLogSegments);
+        // fetch valid data size from BlobStoreStats
+        BlobStoreStats blobStoreStats = blobStore.getBlobStoreStats();
+        Pair<Long, NavigableMap<String, Long>> validDataPerLogSegments = blobStoreStats.getValidDataSizeByLogSegment(
+            new TimeRange(time.milliseconds() - messageRetentionTimeInMs, 0));
+
+        // remove those log segments having overlapping entries in journal
+        List<String> invalidLogSegments = new ArrayList<>();
+        for (String segmentName : validDataPerLogSegments.getSecond().keySet()) {
+          if (!potentialLogSegments.contains(segmentName)) {
+            invalidLogSegments.add(segmentName);
+          }
+        }
+        for (String invalidLogSegment : invalidLogSegments) {
+          validDataPerLogSegments.getSecond().remove(invalidLogSegment);
+        }
+
+        // find best candidate to compact
+        Map.Entry<String, Long> firstEntry = validDataPerLogSegments.getSecond().firstEntry();
+        Map.Entry<String, Long> lastEntry = validDataPerLogSegments.getSecond().lastEntry();
+        CostBenefitInfo bestCandidateToCompact = null;
+        while (!firstEntry.getKey().equals(lastEntry.getKey())) {
```

Comment:
is the ending condition first == last or first > last? first == last seems like a valid case for calculating cost benefit right? That is case where a single log segment may have no valid entries at all and is eligible to be completely dropped. That might be the best cost benefit we get (> 1 and 0 cost).